File: src/org/telegram/messenger/AccountInstance.ts
/**
 * Telegram Official AccountInstance Engine
 * Replicates org.telegram.messenger.AccountInstance.java from DrKLO/Telegram
 * Encapsulates isolated controllers per account (0..MAX_ACCOUNT_COUNT-1).
 */

import { UserConfig } from './UserConfig';
import { MessagesController } from './MessagesController';
import { MessagesStorage } from './MessagesStorage';
import { NotificationCenter } from './NotificationCenter';
import { ConnectionsManager } from '../tgnet/ConnectionsManager';

export class AccountInstance {
  private static instances: AccountInstance[] = [];
  private readonly currentAccount: number;

  private constructor(account: number) {
    this.currentAccount = account;
  }

  public static getInstance(account: number = UserConfig.selectedAccount): AccountInstance {
    if (account < 0 || account >= UserConfig.MAX_ACCOUNT_COUNT) {
      account = 0;
    }
    if (!AccountInstance.instances[account]) {
      AccountInstance.instances[account] = new AccountInstance(account);
    }
    return AccountInstance.instances[account];
  }

  public getCurrentAccount(): number {
    return this.currentAccount;
  }

  public getUserConfig(): UserConfig {
    return UserConfig.getInstance(this.currentAccount);
  }

  public getMessagesController(): MessagesController {
    return MessagesController.getInstance(this.currentAccount);
  }

  public getMessagesStorage(): MessagesStorage {
    return MessagesStorage.getInstance(this.currentAccount);
  }

  public getConnectionsManager(): ConnectionsManager {
    return ConnectionsManager.getInstance(this.currentAccount);
  }

  public getNotificationCenter(): NotificationCenter {
    return NotificationCenter.getInstance(this.currentAccount);
  }
}

File: TMessagesProject/src/main/java/org/telegram/messenger/AccountInstance.java
package org.telegram.messenger;

import org.telegram.tgnet.ConnectionsManager;

/**
 * AccountInstance - Replicates DrKLO/Telegram AccountInstance.java
 * Provides unified isolated access to per-account services:
 * - MessagesController
 * - MessagesStorage
 * - ConnectionsManager
 * - NotificationCenter
 * - UserConfig
 */
public class AccountInstance {
    private final int currentAccount;
    private static final AccountInstance[] Instance = new AccountInstance[UserConfig.MAX_ACCOUNT_COUNT];

    public static AccountInstance getInstance(int num) {
        if (num < 0 || num >= UserConfig.MAX_ACCOUNT_COUNT) {
            num = 0;
        }
        AccountInstance local = Instance[num];
        if (local == null) {
            synchronized (AccountInstance.class) {
                local = Instance[num];
                if (local == null) {
                    Instance[num] = local = new AccountInstance(num);
                }
            }
        }
        return local;
    }

    private AccountInstance(int account) {
        currentAccount = account;
    }

    public int getCurrentAccount() {
        return currentAccount;
    }

    public UserConfig getUserConfig() {
        return UserConfig.getInstance(currentAccount);
    }

    public MessagesController getMessagesController() {
        return MessagesController.getInstance(currentAccount);
    }

    public MessagesStorage getMessagesStorage() {
        return MessagesStorage.getInstance(currentAccount);
    }

    public ConnectionsManager getConnectionsManager() {
        return ConnectionsManager.getInstance(currentAccount);
    }

    public NotificationCenter getNotificationCenter() {
        return NotificationCenter.getInstance(currentAccount);
    }
}
